### PR TITLE
feat(probe): Adding the probe for the chaosexperiments

### PR DESCRIPTION
**Pull request by ispeakc0de**
## Problem Statement
 
As we have generic chaos (supported all apps) so it doesn't contain the application-specific checks. The application-specific checks may vary application to application or setup based. What if someone wants to have application-specific checks?. How will he perform his custom checks? 
- one way is to write separate chaos experiment for each application. sounds complex?. no one is interested to write separate chaos for each application.
- another/ better way is to provide extra power to the user so that he can define his custom checks inside chaos engine CR and the generic experiment can run those checks for him! sounds interesting. Here probes come into play. 

## Solution(Probe)
- Probes can be defined inside the chaos engine and can be trigger inside the chaos experiment and help the user to generate the resilience score.

#### probes defined inside chaosengine

```
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: probe
spec:
  engineStatus: "active"
  auxiliaryAppInfo: "shubham:run=nginx"
  annotationCheck: 'false'
  components:
    runner:
      image: "shubh214/chaos-runner:ci"
      imagePullPolicy: "Always"
  jobCleanUpPolicy: 'retain'
  appinfo: 
    appns: shubham
    applabel: "run=nginx"
    appkind: deployment
  chaosServiceAccount: litmus
  monitoring: false
  experiments:
    - name: pod-delete
      spec:
        # continuous always
        httpProbe:
        - name: "probe-http-1"
          inputs:
            url: "http://10.12.15.53:80"
            expectedResponseCode: "200"
          mode: "EOT"
          runProperties:
            probeTimeout: 10
            interval: 5
            retry: 2
        cmdProbe:
        - name: "probe-cmd-1"
          inputs:
            command: "echo $(cqlsh 10.12.0.97 9042 -e \"DESCRIBE keyspaces\") | grep system_traces | awk '{ print $1}'"
            expectedResult: "system_traces"
            source: "cassandra"
          mode: "SOT"
          runProperties:
            probeTimeout: 10
            interval: 5
            retry: 2
        - name: "probe-cmd-2"
          inputs:
            command: "kubectl get deploy -l run=nginx | grep nginx | awk '{print $1}'"
            expectedResult: "nginx"
            source: "inline"
          mode: "SOT"
          runProperties:
            probeTimeout: 15
            interval: 5
            retry: 1
        k8sProbe:
        - name: "probe-k8s-1"
          inputs:
            command:
              group: ""
              version: "v1"
              resource: "pods"
              namespace: "default"
              fieldSelector: "metadata.name=my-deployment-78ff9649f5-8kmjw,status.phase=Running"
            expectedResult: "shubham"
          mode: "SOT"
          runProperties:
            probeTimeout: 5
            interval: 5
            retry: 2
```

### Types of Probes
1. **Cmd probe:** It can be inline (where any command can be run which supports by our `go-runner` image. OR we can provide source image as well for the other/external dependencies i.e, cassandra/kafka.
2. **HTTP probe:** This probe can be used to get the response from a URL 
3. **K8s probe:** It contains all the checks which need client-go(gvr).

### Modes of the probes
- The mode can be SOT(start of test), EOT(end of test), Edge( SOT and EOT both), or continuous (it will be added soon).

### K8s Probe
It needs a group, version, resource, namespace, and field selector to get the resource using dynamic client & match the result with the expected result.

```
- name: "probe-k8s-1"
          inputs:
            command:
              group: ""
              version: "v1"
              resource: "pods"
              namespace: "default"
              fieldSelector: "metadata.name=my-deployment-78ff9649f5-8kmjw,status.phase=Running"
            expectedResult: "shubham"
          mode: "SOT"
          runProperties:
            probeTimeout: 5
            interval: 5
            retry: 2
```

### Http Probe
It checking for the response status code for a URL. It sends an HTTP request & matches the status code with expected status code.
```
 httpProbe:
        - name: "probe-http-1"
          inputs:
            url: "http://10.12.15.53:80"
            expectedResponseCode: "20"
          mode: "EOT"
          runProperties:
            probeTimeout: 10
            interval: 5
            retry: 2
```

### Cmd Probe
1. It is using cassandra image as source image and run cqlsh command to get the result & match the result w/ expected result

```
cmdProbe:
        - name: "probe-cmd-1"
          inputs:
            command: "echo $(cqlsh 10.12.0.97 9042 -e \"DESCRIBE keyspaces\") | grep system_traces | awk '{ print $1}'"
            expectedResult: "system_traces"
            source: "cassandra"
          mode: "SOT"
          runProperties:
            probeTimeout: 10
            interval: 5
            retry: 2
```
2.  It is inline cmd probe. It is using kubectl commands and match the result with expected result.

```
- name: "probe-cmd-2"
          inputs:
            command: "kubectl get deploy -l run=nginx | grep nginx | awk '{print $1}'"
            expectedResult: "nginx"
            source: "inline"
          mode: "SOT"
          runProperties:
            probeTimeout: 15
            interval: 5
            retry: 1
```

### CODE REVIEW

**Comment by ispeakc0de** [comment]
## CASE-1, When All Probe Passed

###ChaosEngine

```
```
###ChaosResult

```
```
### Logs

```
```

**Comment by ispeakc0de** [comment]
## CASE-2, When Probes Failed in SOT 

### ChaosEngine

```
```
### ChaosResult

```
```
### Logs

```
```

**Comment by ispeakc0de** [comment]
## CASE-3, When Probes Failed In EOT 

### ChaosEngine

```
```
### ChaosResult

```
```
### Logs

```
```

File: experiments/generic/pod-delete/pod-delete.go [hunk 1]
```diff
@@ -75,8 +79,25 @@ func main() {
 		result.RecordAfterFailure(&chaosDetails, &resultDetails, failStep, clients, &eventsDetails)
 		return
 	}
+
 	if experimentsDetails.EngineName != "" {
-		types.SetEngineEventAttributes(&eventsDetails, types.PreChaosCheck, "AUT is Running successfully", "Normal", &chaosDetails)
+
+		// Add the probes in the pre-chaos check
+		err = probe.AddProbes(&chaosDetails, clients, &resultDetails, "PreChaos", &eventsDetails)
```
Comment (ksatchit) [inline experiments/generic/pod-delete/pod-delete.go:86]: Can we say RunProbe() instead

File: pkg/probe/k8sprobe.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+			if (probe.Mode == "SOT" && phase == "PreChaos") || (probe.Mode == "EOT" && phase == "PostChaos") || probe.Mode == "Edge" {
+
+				//DISPLAY THE K8S PROBE INFO
+				log.InfoWithValues("[Probe]: The k8s probe informations are as follows", logrus.Fields{
+					"Name":            probe.Name,
+					"Command":         probe.Inputs.Command,
+					"Expected Result": probe.Inputs.ExpectedResult,
+					"Run Properties":  probe.RunProperties,
+					"Mode":            probe.Mode,
+				})
```
Comment (ksatchit) [inline pkg/probe/k8sprobe.go:36]: 👍

File: pkg/probe/probe.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			cmdProbes = experiment.Spec.CmdProbe
+		}
+	}
+
+	return k8sProbes, cmdProbes, httpProbes, nil
+}
+
+// SetProbesInChaosResult set the probe inside chaos result
+// it fetch the probe details from the chaosengine and set into the chaosresult
+func SetProbesInChaosResult(chaosDetails *types.ChaosDetails, clients clients.ClientSets, chaosresult *types.ResultDetails) error {
```
Comment (ksatchit) [inline pkg/probe/probe.go:104]: InitializeProbesInChaosResultDetails

File: pkg/probe/httpprobe.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		TryWithTimeout(func(attempt uint) error {
+			// getting the response from the given url
+			resp, err := http.Get(probe.Inputs.URL)
+			if err != nil {
+				return err
+			}
+			code, _ := strconv.Atoi(probe.Inputs.ExpectedResponseCode)
+			// matching the status code w/ expected code
+			if resp.StatusCode != code {
+				return fmt.Errorf("not getting 200 status code, code: %v", resp.StatusCode)
```
Comment (ksatchit) [inline pkg/probe/httpprobe.go:92]: expected status code over "200"

File: pkg/result/chaosresult.go [hunk 5]
```diff
@@ -108,13 +112,47 @@ func InitializeChaosResult(chaosDetails *types.ChaosDetails, clients clients.Cli
...
 	result.Spec.InstanceID = chaosDetails.InstanceID
 	result.Status.ExperimentStatus.FailStep = resultDetails.FailStep
+	result.Status.ProbeStatus = GetProbeStatus(resultDetails)
+	if resultDetails.Phase == "Completed" {
+		if resultDetails.Verdict == "Pass" {
+			if len(resultDetails.ProbeDetails) == 0 {
+				result.Status.ExperimentStatus.ResilienceScore = "100%"
+			} else {
+
+				result.Status.ExperimentStatus.ResilienceScore = strconv.Itoa((resultDetails.ProbeCount*60)/len(resultDetails.ProbeDetails)+40) + "%"
```
Comment (ksatchit) [inline pkg/result/chaosresult.go:143]: result.Status.ExperimentStatus.probeSuccessPercentage = passedProbeCount/TotalProbeCount

File: pkg/probe/cmdprobe.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+			"ProbeInstance": phase,
+			"ProbeStatus":   "Fail",
+		})
+		SetProbeVerdictAfterFailure(resultDetails)
+		return err
+	}
+	// counting the passed probes count to generate the score and mark the verdict as passed
+	// for edge, probe is marked as Passed if passed in both pre/post chaos checks
+	if !(probe.Mode == "Edge" && phase == "PreChaos") {
+		resultDetails.ProbeCount++
```
Comment (ksatchit) [inline pkg/probe/cmdprobe.go:247]: passedProbeCount

File: pkg/result/chaosresult.go [hunk 7]
```diff
@@ -108,13 +112,41 @@ func InitializeChaosResult(chaosDetails *types.ChaosDetails, clients clients.Cli
...
 
 	result.Status.ExperimentStatus.Phase = resultDetails.Phase
 	result.Status.ExperimentStatus.Verdict = resultDetails.Verdict
 	result.Spec.InstanceID = chaosDetails.InstanceID
 	result.Status.ExperimentStatus.FailStep = resultDetails.FailStep
+	result.Status.ProbeStatus = GetProbeStatus(resultDetails)
+	if resultDetails.Phase == "Completed" {
+		if resultDetails.Verdict == "Pass" && len(resultDetails.ProbeDetails) != 0 {
+
+			result.Status.ExperimentStatus.ProbeSuccessPercentage = "100%"
```
Comment (ksatchit) [inline pkg/result/chaosresult.go:140]: can we remove the `%` char - that we way can consume directly in metric

**Comment by ispeakc0de** [comment]
## After final changes

### Logs

```
time="2020-08-13T09:52:14Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2020-08-13T09:52:14Z" level=info msg="[PreReq]: Updating the chaos result of pod-delete experiment (SOT)"
time="2020-08-13T09:52:14Z" level=info msg="The application informations are as follows" Namespace=shubham Label="run=nginx" Ramp Time=0
time="2020-08-13T09:52:14Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2020-08-13T09:52:14Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-13T09:52:14Z" level=info msg="The running status of Pods are as follows" Pod=nginx-6876f9c9f8-sgjnk Status=Running
time="2020-08-13T09:52:14Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-13T09:52:14Z" level=info msg="The running status of container are as follows" Pod=nginx-6876f9c9f8-sgjnk Status=Running container=nginx
time="2020-08-13T09:52:14Z" level=info msg="[Probe]: The k8s probe informations are as follows" Name=probe-k8s-1 Command="{ v1 pods default metadata.name=my-deployment-78ff9649f5-2vjpx,status.phase=Running}" Expected Result=shubham Run Properties="{5 5 2}" Mode=SOT
time="2020-08-13T09:52:19Z" level=info msg="[Probe]: k8s probe has been Passed 😄 " ProbeStatus=Pass ProbeName=probe-k8s-1 ProbeType=K8sProbe ProbeInstance=PreChaos
time="2020-08-13T09:52:19Z" level=info msg="[Probe]: The cmd probe informations are as follows" Mode=SOT Name=probe-cmd-1 Command="echo $(cqlsh 10.12.0.183 9042 -e \"DESCRIBE keyspaces\") | grep system_schema | awk '{ print $1}'" Expected Result=system_traces Source=cassandra Run Properties="{10 5 2}"
time="2020-08-13T09:52:19Z" level=info msg="[Status]: Checking the status of the probe pod"
time="2020-08-13T09:52:19Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-13T09:52:23Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=pod-delete-probe-elmjfd
time="2020-08-13T09:52:23Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-13T09:52:23Z" level=info msg="The running status of container are as follows" Status=Running container=pod-delete-probe Pod=pod-delete-probe-elmjfd
time="2020-08-13T09:52:29Z" level=info msg="[Probe]: cmd probe has been Passed 😄 " ProbeName=probe-cmd-1 ProbeType=CmdProbe ProbeInstance=PreChaos ProbeStatus=Pass
time="2020-08-13T09:52:31Z" level=info msg="[Probe]: The cmd probe informations are as follows" Run Properties="{15 5 1}" Mode=SOT Name=probe-cmd-2 Command="kubectl get pods -l run=nginx -o jsonpath='{.items[*].spec.containers[?(@.name==\"nginx\")].resources.limits.ephemeral-storage}' | grep -Eo '[[:alpha:]]+'" Expected Result=Gi Source=inline
time="2020-08-13T09:52:37Z" level=info msg="[Probe]: cmd probe has been Passed 😄 " ProbeType=CmdProbe ProbeInstance=PreChaos ProbeStatus=Pass ProbeName=probe-cmd-2
time="2020-08-13T09:52:37Z" level=info msg="[Info]: Killing the following pods" PodList="[nginx-6876f9c9f8-sgjnk]"
time="2020-08-13T09:52:37Z" level=info msg="[Wait]: Wait for the chaos interval 10s"
time="2020-08-13T09:52:47Z" level=info msg="[Status]: Verification for the recreation of application pod"
time="2020-08-13T09:52:47Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-13T09:52:47Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=nginx-6876f9c9f8-8zxrh
time="2020-08-13T09:52:47Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-13T09:52:47Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-6876f9c9f8-8zxrh Status=Running
time="2020-08-13T09:52:47Z" level=info msg="[Info]: Killing the following pods" PodList="[nginx-6876f9c9f8-8zxrh]"
time="2020-08-13T09:52:47Z" level=info msg="[Wait]: Wait for the chaos interval 10s"
time="2020-08-13T09:52:57Z" level=info msg="[Status]: Verification for the recreation of application pod"
time="2020-08-13T09:52:57Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-13T09:52:57Z" level=info msg="The running status of Pods are as follows" Pod=nginx-6876f9c9f8-krld7 Status=Running
time="2020-08-13T09:52:57Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-13T09:52:57Z" level=info msg="The running status of container are as follows" Pod=nginx-6876f9c9f8-krld7 Status=Running container=nginx
time="2020-08-13T09:52:57Z" level=info msg="[Info]: Killing the following pods" PodList="[nginx-6876f9c9f8-krld7]"
time="2020-08-13T09:52:57Z" level=info msg="[Wait]: Wait for the chaos interval 10s"
time="2020-08-13T09:53:07Z" level=info msg="[Status]: Verification for the recreation of application pod"
time="2020-08-13T09:53:07Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-13T09:53:07Z" level=info msg="The running status of Pods are as follows" Pod=nginx-6876f9c9f8-5qtx6 Status=Running
time="2020-08-13T09:53:07Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-13T09:53:07Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-6876f9c9f8-5qtx6 Status=Running
time="2020-08-13T09:53:07Z" level=info msg="[Completion]: pod-delete chaos is done"
time="2020-08-13T09:53:07Z" level=info msg="[Confirmation]: The application pod has been deleted successfully"
time="2020-08-13T09:53:07Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2020-08-13T09:53:07Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-13T09:53:07Z" level=info msg="The running status of Pods are as follows" Pod=nginx-6876f9c9f8-5qtx6 Status=Running
time="2020-08-13T09:53:07Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-13T09:53:07Z" level=info msg="The running status of container are as follows" Status=Running container=nginx Pod=nginx-6876f9c9f8-5qtx6
time="2020-08-13T09:53:07Z" level=info msg="[Probe]: The http probe informations are as follows" Expecected Response Code=200 Run Properties="{10 5 2}" Mode=EOT Name=probe-http-1 URL="http://10.12.0.75:80"
time="2020-08-13T09:53:12Z" level=info msg="[Probe]: http probe has been Passed 😄 " ProbeStatus=Pass ProbeName=probe-http-1 ProbeType=HTTPProbe ProbeInstance=PostChaos
time="2020-08-13T09:53:12Z" level=info msg="[The End]: Updating the chaos result of pod-delete experiment (EOT)"

```

### Engine

```
Name:         probe
Namespace:    shubham
Labels:       <none>
Annotations:  <none>
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosEngine
Metadata:
  Creation Timestamp:  2020-08-13T09:51:37Z
  Finalizers:
    chaosengine.litmuschaos.io/finalizer
  Generation:        22
  Resource Version:  57111
  Self Link:         /apis/litmuschaos.io/v1alpha1/namespaces/shubham/chaosengines/probe
  UID:               df69b860-fe72-4e60-9079-325aa27a14a6
Spec:
  Annotation Check:  false
  Appinfo:
    Appkind:              deployment
    Applabel:             run=nginx
    Appns:                shubham
  Auxiliary App Info:     shubham:run=nginx
  Chaos Service Account:  litmus
  Components:
    Runner:
      Image:              shubh214/chaos-runner:ci
      Image Pull Policy:  Always
  Engine State:           stop
  Experiments:
    Name:  pod-delete
    Spec:
      Cmd Probe:
        Inputs:
          Command:          echo $(cqlsh 10.12.0.183 9042 -e "DESCRIBE keyspaces") | grep system_schema | awk '{ print $1}'
          Expected Result:  system_traces
          Source:           cassandra
        Mode:               SOT
        Name:               probe-cmd-1
        Run Properties:
          Interval:       5
          Probe Timeout:  10
          Retry:          2
        Inputs:
          Command:          kubectl get pods -l run=nginx -o jsonpath='{.items[*].spec.containers[?(@.name=="nginx")].resources.limits.ephemeral-storage}' | grep -Eo '[[:alpha:]]+'
          Expected Result:  Gi
          Source:           inline
        Mode:               SOT
        Name:               probe-cmd-2
        Run Properties:
          Interval:       5
          Probe Timeout:  15
          Retry:          1
      Components:
      Http Probe:
        Inputs:
          Expected Response Code:  200
          URL:                     http://10.12.0.75:80
        Mode:                      EOT
        Name:                      probe-http-1
        Run Properties:
          Interval:       5
          Probe Timeout:  10
          Retry:          2
      k8sProbe:
        Inputs:
          Command:
            Field Selector:  metadata.name=my-deployment-78ff9649f5-2vjpx,status.phase=Running
            Namespace:       default
            Resource:        pods
            Version:         v1
          Expected Result:   shubham
        Mode:                SOT
        Name:                probe-k8s-1
        Run Properties:
          Interval:       5
          Probe Timeout:  5
          Retry:          2
      Rank:               0
  Job Clean Up Policy:    retain
Status:
  Engine Status:  completed
  Experiments:
    Last Update Time:  2020-08-13T09:53:19Z
    Name:              pod-delete-zqos45
    Status:            Execution Successful
    Verdict:           Pass
Events:
  Type    Reason                     Age                From                     Message
  ----    ------                     ----               ----                     -------
  Normal  ChaosEngineInitialized     2m28s              chaos-operator           probe-runner created successfully
  Normal  ExperimentDependencyCheck  2m22s              probe-runner             Experiment resources validated for Chaos Experiment: 'pod-delete'
  Normal  ExperimentJobCreate        2m17s              probe-runner             Experiment Job 'pod-delete-zqos45' created for Chaos Experiment 'pod-delete'
  Normal  PreChaosCheck              88s                pod-delete-zqos45-x7pd2  AUT: Running, Probes: Successful
  Normal  ChaosInject                68s (x3 over 88s)  pod-delete-zqos45-x7pd2  Injecting pod-delete chaos on application pod
  Normal  PostChaosCheck             53s                pod-delete-zqos45-x7pd2  AUT: Running, Probes: Successful
  Normal  Summary                    53s                pod-delete-zqos45-x7pd2  pod-delete experiment has been Passed
  Normal  ExperimentJobCleanUp       46s                probe-runner             Experiment Job 'pod-delete-zqos45' will be retained
  Normal  ChaosEngineCompleted       40s                chaos-operator           Chaos Engine completed, will delete or retain the resources according to jobCleanUpPolicy

```

### Result

```
Name:         probe-pod-delete
Namespace:    shubham
Labels:       name=probe-pod-delete
Annotations:  <none>
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosResult
Metadata:
  Creation Timestamp:  2020-08-13T09:52:14Z
  Generation:          2
  Resource Version:    57058
  Self Link:           /apis/litmuschaos.io/v1alpha1/namespaces/shubham/chaosresults/probe-pod-delete
  UID:                 2925fab6-a81a-4a74-9faa-d7bda2cab59e
Spec:
  Engine:      probe
  Experiment:  pod-delete
Status:
  Experimentstatus:
    Fail Step:                 N/A
    Phase:                     Completed
    Probe Success Percentage:  100
    Verdict:                   Pass
  Probe Status:
    Name:  probe-k8s-1
    Status:
      Pre Chaos:  Passed 👍 
    Type:         K8sProbe
    Name:         probe-http-1
    Status:
      Post Chaos:  Passed 👍 
    Type:          HTTPProbe
    Name:          probe-cmd-1
    Status:
      Pre Chaos:  Passed 👍 
    Type:         CmdProbe
    Name:         probe-cmd-2
    Status:
      Pre Chaos:  Passed 👍 
    Type:         CmdProbe
Events:
  Type    Reason   Age   From                     Message
  ----    ------   ----  ----                     -------
  Normal  Summary  86s   pod-delete-zqos45-x7pd2  pod-delete experiment has been Passed

```

File: experiments/generic/pod-delete/pod-delete.go [hunk 8]
```diff
@@ -75,8 +79,25 @@ func main() {
...
 	}
+
 	if experimentsDetails.EngineName != "" {
-		types.SetEngineEventAttributes(&eventsDetails, types.PreChaosCheck, "AUT is Running successfully", "Normal", &chaosDetails)
+
+		// run the probes in the pre-chaos check
+		err = probe.RunProbes(&chaosDetails, clients, &resultDetails, "PreChaos", &eventsDetails)
+		events.GenerateEvents(&eventsDetails, clients, &chaosDetails, "ChaosEngine")
+		if err != nil {
+			log.Errorf("Unable to Add the probes, due to err: %v", err)
```
Comment (ksatchit) [inline experiments/generic/pod-delete/pod-delete.go:89]: Lets call it: `Probe failed` and have this log or the previous one mention the name of the probe.

**Pull request merged**